experimental/schoology/export-assignment-data.py

```diff
@@ -0,0 +1,269 @@
+"""
+This script retrieves and writes out to CSV all students, sections, assignments,
+and submissions for the desired grading period(s). Configuration is via `.env`
+file or directly with environment variables:
+
+* SCHOOLOGY_KEY=<administrator's API key>
+* SCHOOLOGY_SECRET=<administrator's API secret>
+* SCHOOLOGY_GRADING_PERIODS=<csv list of grading period ids>
+"""
+
+from datetime import datetime
+import os
+
+from dotenv import load_dotenv
+import pandas as pd
+import schoolopy
+from schoolopy.models import Submission
+
+
+class SchoologyExport:
+    api_key = ""
+    api_secret = ""
+    desired_grading_periods = []
+    grading_periods_lookup = {}
+    output_path = ""
+    sc_client = None
+
+    def __init__(self):
+        # TODO: find a better way. This value is set for Ed-Fi/MSDF users
+        # because of an internal network issue. Necessary for running PIP. But
+        # it messes with the commands below, so needs to be unset.
+        os.environ["REQUESTS_CA_BUNDLE"] = ""
+
+        self._get_configuration()
+        self._initialize_connection()
+        self._get_grading_periods()
+
+    def _get_configuration(self):
+        # TODO: long-term, probably want to allow values to be set at the command
+        # line as alternate to environment variables.
+
+        load_dotenv()
+
+        ASSERT_TEMPLATE = "{key} environment variable is not set."
+
+        self.api_key = os.getenv("SCHOOLOGY_KEY")
+        assert self.api_key is not None, ASSERT_TEMPLATE.format("SCHOOLOGY_KEY")
+
+        self.api_secret = os.getenv("SCHOOLOGY_SECRET")
+        assert self.api_secret is not None, ASSERT_TEMPLATE.format("SCHOOLOGY_SECRET")
+
+        self.output_path = os.getenv("SCHOOLOGY_OUTPUT_PATH")
+        assert self.api_secret is not None, ASSERT_TEMPLATE.format(
+            "SCHOOLOGY_OUTPUT_PATH"
+        )
+
+        desired_grading_periods = os.getenv("SCHOOLOGY_GRADING_PERIODS")
+        assert desired_grading_periods is not None, ASSERT_TEMPLATE.format(
+            "SCHOOLOGY_GRADING_PERIODS"
+        )
+
+        try:
+            self.desired_grading_periods = [
+                int(gp) for gp in desired_grading_periods.split(",")
+            ]
+        except ValueError as ex:
+            msg = f"Grading period environment variable could not be parsed: {ex}"
+            raise Exception(msg)
+
+    def _initialize_connection(self):
+        self.sc_client = schoolopy.Schoology(
+            schoolopy.Auth(self.api_key, self.api_secret)
+        )
+
+    def _get_grading_periods(self):
+        gp_list = self.sc_client.get_grading_periods()
+        gl = {
+            gp.id: gp.title for gp in gp_list if gp.id in self.desired_grading_periods
+        }
+        self.grading_periods_lookup = gl

```

Comment:
We only want data "for the current school year." There is no native concept of that in Schoology. So there are grading_period_ids in the environment variables to identify the periods of interest. I intended to include a small script to help someone look up the correct IDs but didn't get around to it yet. This function looks up all of the grading periods in the API and then filters to have only the desired ones, storing them in a lookup table for reference later on.